File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImageUtilTests.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "ImagePickerTestImages.h"

@import image_picker_ios;
#if __has_include(<image_picker_ios/image_picker_ios-umbrella.h>)
@import image_picker_ios.Test;
#endif
@import XCTest;

// Corner colors of test image scaled to 3x2. Format is "R G B A".
static NSString *const kColorRepresentation3x2BottomLeftYellow = @"1 0.776471 0 1";
static NSString *const kColorRepresentation3x2TopLeftRed = @"1 0.0666667 0 1";
static NSString *const kColorRepresentation3x2BottomRightCyan = @"0 0.772549 1 1";
static NSString *const kColorRepresentation3x2TopRightBlue = @"0 0.0705882 0.996078 1";

@interface ImageUtilTests : XCTestCase
@end

@implementation ImageUtilTests

static NSString *ColorStringAtPixel(UIImage *image, int pixelX, int pixelY) {
  CGImageRef cgImage = image.CGImage;

  uint32_t argb;
  CGContextRef context1 = CGBitmapContextCreate(
      &argb, 1, 1, CGImageGetBitsPerComponent(cgImage), CGImageGetBytesPerRow(cgImage),
      CGColorSpaceCreateDeviceRGB(), CGImageGetBitmapInfo(cgImage));
  CGContextDrawImage(
      context1, CGRectMake(-pixelX, -pixelY, CGImageGetWidth(cgImage), CGImageGetHeight(cgImage)),
      cgImage);
  CGContextRelease(context1);
  int blue = argb & 0xff;
  int green = argb >> 8 & 0xff;
  int red = argb >> 16 & 0xff;
  int alpha = argb >> 24 & 0xff;

  return [CIColor colorWithRed:red / 255.f
                         green:green / 255.f
                          blue:blue / 255.f
                         alpha:alpha / 255.f]
      .stringRepresentation;
}

- (void)testScaledImage_EqualSizeReturnsSameImage {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  UIImage *scaledImage = [FLTImagePickerImageUtil scaledImage:image
                                                     maxWidth:@(image.size.width)
                                                    maxHeight:@(image.size.height)
                                          isMetadataAvailable:YES];

  // Assert the same bytes pointer (not just equal objects).
  XCTAssertEqual(image, scaledImage);
}

- (void)testScaledImage_NilSizeReturnsSameImage {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  UIImage *scaledImage = [FLTImagePickerImageUtil scaledImage:image
                                                     maxWidth:nil
                                                    maxHeight:nil
                                          isMetadataAvailable:YES];

  // Assert the same bytes pointer (not just equal objects).
  XCTAssertEqual(image, scaledImage);
}

- (void)testScaledImage_ShouldBeScaled {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];

  CGFloat scaledWidth = 3;
  CGFloat scaledHeight = 2;
  UIImage *scaledImage = [FLTImagePickerImageUtil scaledImage:image
                                                     maxWidth:@(scaledWidth)
                                                    maxHeight:@(scaledHeight)
                                          isMetadataAvailable:YES];
  XCTAssertEqual(scaledImage.size.width, scaledWidth);
  XCTAssertEqual(scaledImage.size.height, scaledHeight);

  // Check the corners to make sure nothing has been rotated.
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, 0, 0),
                        kColorRepresentation3x2BottomLeftYellow);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, 0, scaledHeight - 1),
                        kColorRepresentation3x2TopLeftRed);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, scaledWidth - 1, 0),
                        kColorRepresentation3x2BottomRightCyan);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, scaledWidth - 1, scaledHeight - 1),
                        kColorRepresentation3x2TopRightBlue);
}

- (void)testScaledImage_ShouldBeScaledWithNoMetadata {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];

  CGFloat scaledWidth = 3;
  CGFloat scaledHeight = 2;
  UIImage *scaledImage = [FLTImagePickerImageUtil scaledImage:image
                                                     maxWidth:@(scaledWidth)
                                                    maxHeight:@(scaledHeight)
                                          isMetadataAvailable:NO];
  XCTAssertEqual(scaledImage.size.width, scaledWidth);
  XCTAssertEqual(scaledImage.size.height, scaledHeight);

  // Check the corners to make sure nothing has been rotated.
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, 0, 0),
                        kColorRepresentation3x2BottomLeftYellow);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, 0, scaledHeight - 1),
                        kColorRepresentation3x2TopLeftRed);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, scaledWidth - 1, 0),
                        kColorRepresentation3x2BottomRightCyan);
  XCTAssertEqualObjects(ColorStringAtPixel(scaledImage, scaledWidth - 1, scaledHeight - 1),
                        kColorRepresentation3x2TopRightBlue);
}

- (void)testScaledImage_ShouldBeCorrectRotation {
  NSURL *imageURL =
      [[NSBundle bundleForClass:[self class]] URLForResource:@"jpgImageWithRightOrientation"
                                               withExtension:@"jpg"];
  NSData *imageData = [NSData dataWithContentsOfURL:imageURL];
  UIImage *image = [UIImage imageWithData:imageData];
  XCTAssertEqual(image.size.width, 130);
  XCTAssertEqual(image.size.height, 174);
  XCTAssertEqual(image.imageOrientation, UIImageOrientationRight);

  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@10
                                                 maxHeight:@10
                                       isMetadataAvailable:YES];
  XCTAssertEqual(newImage.size.width, 10);
  XCTAssertEqual(newImage.size.height, 7);
  XCTAssertEqual(newImage.imageOrientation, UIImageOrientationUp);
}

- (void)testScaledGIFImage_ShouldBeScaled {
  // gif image that frame size is 3 and the duration is 1 second.
  GIFInfo *info = [FLTImagePickerImageUtil scaledGIFImage:ImagePickerTestImages.GIFTestData
                                                 maxWidth:@3
                                                maxHeight:@2];

  NSArray<UIImage *> *images = info.images;
  NSTimeInterval duration = info.interval;

  XCTAssertEqual(images.count, 3);
  XCTAssertEqual(duration, 1);

  for (UIImage *newImage in images) {
    XCTAssertEqual(newImage.size.width, 3);
    XCTAssertEqual(newImage.size.height, 2);
  }
}

- (void)testScaledImage_TallImage_ShouldBeScaledBelowMaxHeight {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTallTestData];
  XCTAssertEqual(image.size.width, 4);
  XCTAssertEqual(image.size.height, 7);
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@5
                                                 maxHeight:@5
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 3);
  XCTAssertEqual(newImage.size.height, 5);
}

- (void)testScaledImage_TallImage_ShouldBeScaledBelowMaxWidth {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTallTestData];
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@3
                                                 maxHeight:@10
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 3);
  XCTAssertEqual(newImage.size.height, 5);
}

- (void)testScaledImage_TallImage_ShouldNotBeScaledAboveOriginaWidthOrHeight {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTallTestData];
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@10
                                                 maxHeight:@10
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 4);
  XCTAssertEqual(newImage.size.height, 7);
}

- (void)testScaledImage_WideImage_ShouldBeScaledBelowMaxHeight {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  XCTAssertEqual(image.size.width, 12);
  XCTAssertEqual(image.size.height, 7);
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@20
                                                 maxHeight:@6
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 10);
  XCTAssertEqual(newImage.size.height, 6);
}

- (void)testScaledImage_WideImage_ShouldBeScaledBelowMaxWidth {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@10
                                                 maxHeight:@10
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 10);
  XCTAssertEqual(newImage.size.height, 6);
}

- (void)testScaledImage_WideImage_ShouldNotBeScaledAboveOriginaWidthOrHeight {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@100
                                                 maxHeight:@100
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage.size.width, 12);
  XCTAssertEqual(newImage.size.height, 7);
}

- (void)testScaledImage_ImageIsNil {
  UIImage *image = nil;
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@1440
                                                 maxHeight:@1440
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage, nil);
}

- (void)testScaledImage_ImageMaxWidthZeroAndMaxHeightIsZero {
  UIImage *image = [UIImage imageWithData:ImagePickerTestImages.JPGTestData];
  UIImage *newImage = [FLTImagePickerImageUtil scaledImage:image
                                                  maxWidth:@0
                                                 maxHeight:@0
                                       isMetadataAvailable:YES];

  XCTAssertEqual(newImage, nil);
}

@end

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImageUtilTests.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import ImageIO
import Testing
import UIKit
import UniformTypeIdentifiers

@testable import image_picker_ios

private let kColorRepresentation3x2BottomLeftYellow = "1 0.776471 0 1"
private let kColorRepresentation3x2TopLeftRed = "1 0.0666667 0 1"
private let kColorRepresentation3x2BottomRightCyan = "0 0.772549 1 1"
private let kColorRepresentation3x2TopRightBlue = "0 0.0705882 0.996078 1"

private func colorString(atPixel image: UIImage, pixelX: Int, pixelY: Int) -> String {
  let cgImage = image.cgImage!
  var argb: UInt32 = 0
  let context = CGContext(
    data: &argb,
    width: 1,
    height: 1,
    bitsPerComponent: cgImage.bitsPerComponent,
    bytesPerRow: cgImage.bytesPerRow,
    space: CGColorSpaceCreateDeviceRGB(),
    bitmapInfo: cgImage.bitmapInfo.rawValue)!
  context.draw(
    cgImage,
    in: CGRect(
      x: -CGFloat(pixelX), y: -CGFloat(pixelY), width: CGFloat(cgImage.width),
      height: CGFloat(cgImage.height)))
  let blue = Int(argb & 0xff)
  let green = Int(argb >> 8 & 0xff)
  let red = Int(argb >> 16 & 0xff)
  let alpha = Int(argb >> 24 & 0xff)
  return CIColor(
    red: CGFloat(red) / 255.0,
    green: CGFloat(green) / 255.0,
    blue: CGFloat(blue) / 255.0,
    alpha: CGFloat(alpha) / 255.0
  ).stringRepresentation
}

@Suite
struct ImageUtilTests {
  @Test func scaledImageEqualSizeReturnsSameImage() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let scaledImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: NSNumber(value: image.size.width),
      maxHeight: NSNumber(value: image.size.height), isMetadataAvailable: true)
    #expect(image === scaledImage)
  }

  @Test func scaledImageNilSizeReturnsSameImage() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let scaledImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: nil, maxHeight: nil, isMetadataAvailable: true)
    #expect(image === scaledImage)
  }

  @Test func scaledImageShouldBeScaled() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let scaledWidth: CGFloat = 3
    let scaledHeight: CGFloat = 2
    let scaledImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: NSNumber(value: scaledWidth), maxHeight: NSNumber(value: scaledHeight),
      isMetadataAvailable: true)!
    #expect(scaledImage.size.width == scaledWidth)
    #expect(scaledImage.size.height == scaledHeight)
    #expect(
      colorString(atPixel: scaledImage, pixelX: 0, pixelY: 0)
        == kColorRepresentation3x2BottomLeftYellow)
    #expect(
      colorString(atPixel: scaledImage, pixelX: 0, pixelY: Int(scaledHeight) - 1)
        == kColorRepresentation3x2TopLeftRed)
    #expect(
      colorString(atPixel: scaledImage, pixelX: Int(scaledWidth) - 1, pixelY: 0)
        == kColorRepresentation3x2BottomRightCyan)
    #expect(
      colorString(
        atPixel: scaledImage, pixelX: Int(scaledWidth) - 1, pixelY: Int(scaledHeight) - 1)
        == kColorRepresentation3x2TopRightBlue)
  }

  @Test func scaledImageShouldBeScaledWithNoMetadata() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let scaledWidth: CGFloat = 3
    let scaledHeight: CGFloat = 2
    let scaledImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: NSNumber(value: scaledWidth), maxHeight: NSNumber(value: scaledHeight),
      isMetadataAvailable: false)!
    #expect(scaledImage.size.width == scaledWidth)
    #expect(scaledImage.size.height == scaledHeight)
    #expect(
      colorString(atPixel: scaledImage, pixelX: 0, pixelY: 0)
        == kColorRepresentation3x2BottomLeftYellow)
    #expect(
      colorString(atPixel: scaledImage, pixelX: 0, pixelY: Int(scaledHeight) - 1)
        == kColorRepresentation3x2TopLeftRed)
    #expect(
      colorString(atPixel: scaledImage, pixelX: Int(scaledWidth) - 1, pixelY: 0)
        == kColorRepresentation3x2BottomRightCyan)
    #expect(
      colorString(
        atPixel: scaledImage, pixelX: Int(scaledWidth) - 1, pixelY: Int(scaledHeight) - 1)
        == kColorRepresentation3x2TopRightBlue)
  }

  @Test func scaledImageShouldBeCorrectRotation() throws {
    let imageURL = try #require(
      Bundle(for: ImagePickerTestImages.self).url(
        forResource: "jpgImageWithRightOrientation", withExtension: "jpg"))
    let imageData = try Data(contentsOf: imageURL)
    let image = try #require(UIImage(data: imageData))
    #expect(image.size.width == 130)
    #expect(image.size.height == 174)
    #expect(image.imageOrientation == .right)

    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 10, maxHeight: 10, isMetadataAvailable: true)!
    #expect(newImage.size.width == 10)
    #expect(newImage.size.height == 7)
    #expect(newImage.imageOrientation == .up)
  }

  @Test func scaledGIFImageShouldBeScaled() {
    let info = FLTImagePickerImageUtil.scaledGIFImage(
      ImagePickerTestImages.gifTestData, maxWidth: 3, maxHeight: 2)
    #expect(info.images.count == 3)
    #expect(info.interval == 1)
    for newImage in info.images {
      #expect(newImage.size.width == 3)
      #expect(newImage.size.height == 2)
    }
  }

  @Test func scaledImageTallImageShouldBeScaledBelowMaxHeight() {
    let image = UIImage(data: ImagePickerTestImages.jpgTallTestData)!
    #expect(image.size.width == 4)
    #expect(image.size.height == 7)
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 5, maxHeight: 5, isMetadataAvailable: true)!
    #expect(newImage.size.width == 3)
    #expect(newImage.size.height == 5)
  }

  @Test func scaledImageTallImageShouldBeScaledBelowMaxWidth() {
    let image = UIImage(data: ImagePickerTestImages.jpgTallTestData)!
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 3, maxHeight: 10, isMetadataAvailable: true)!
    #expect(newImage.size.width == 3)
    #expect(newImage.size.height == 5)
  }

  @Test func scaledImageTallImageShouldNotBeScaledAboveOriginalWidthOrHeight() {
    let image = UIImage(data: ImagePickerTestImages.jpgTallTestData)!
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 10, maxHeight: 10, isMetadataAvailable: true)!
    #expect(newImage.size.width == 4)
    #expect(newImage.size.height == 7)
  }

  @Test func scaledImageWideImageShouldBeScaledBelowMaxHeight() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    #expect(image.size.width == 12)
    #expect(image.size.height == 7)
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 20, maxHeight: 6, isMetadataAvailable: true)!
    #expect(newImage.size.width == 10)
    #expect(newImage.size.height == 6)
  }

  @Test func scaledImageWideImageShouldBeScaledBelowMaxWidth() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 10, maxHeight: 10, isMetadataAvailable: true)!
    #expect(newImage.size.width == 10)
    #expect(newImage.size.height == 6)
  }

  @Test func scaledImageWideImageShouldNotBeScaledAboveOriginalWidthOrHeight() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 100, maxHeight: 100, isMetadataAvailable: true)!
    #expect(newImage.size.width == 12)
    #expect(newImage.size.height == 7)
  }

  @Test func scaledImageImageIsNil() {
    let newImage = FLTImagePickerImageUtil.scaledImage(
      nil, maxWidth: 1440, maxHeight: 1440, isMetadataAvailable: true)
    #expect(newImage == nil)
  }

  @Test func scaledImageImageMaxWidthZeroAndMaxHeightIsZero() {
    let image = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let newImage = FLTImagePickerImageUtil.scaledImage(
      image, maxWidth: 0, maxHeight: 0, isMetadataAvailable: true)
    #expect(newImage == nil)
  }

  @Test func scaledGIFImageUsesClampedDelayWhenUnclampedDelayMissing() {
    let frame = UIImage(data: ImagePickerTestImages.jpgTestData)!
    let gifData = NSMutableData()
    let destination = CGImageDestinationCreateWithData(
      gifData, UTType.gif.identifier as CFString, 2, nil)!
    let frameProperties: [CFString: Any] = [
      kCGImagePropertyGIFDictionary: [
        kCGImagePropertyGIFDelayTime: 0.25
      ]
    ]
    CGImageDestinationAddImage(destination, frame.cgImage!, frameProperties as CFDictionary)
    CGImageDestinationAddImage(destination, frame.cgImage!, frameProperties as CFDictionary)
    #expect(CGImageDestinationFinalize(destination))

    let info = FLTImagePickerImageUtil.scaledGIFImage(
      gifData as Data, maxWidth: 3, maxHeight: 2)
    #expect(info.images.count == 2)
    #expect(abs(info.interval - 0.25) < 0.001)
  }
}
